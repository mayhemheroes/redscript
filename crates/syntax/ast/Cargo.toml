[package]
name = "redscript-ast"
version.workspace = true
authors.workspace = true
edition.workspace = true

[dependencies]
derive-where.workspace = true
bitflags.workspace = true
elsa.workspace = true
chumsky = { workspace = true, optional = true }
ignore = { workspace = true, optional = true }

[lints]
workspace = true

[features]
chumsky = ["dep:chumsky"]
ignore = ["dep:ignore"]
testing = []
