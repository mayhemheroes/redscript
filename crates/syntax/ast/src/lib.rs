mod ast;
mod files;
mod span;
mod visitor;

pub use ast::{
    Aggregate, Annotation, ArraySpread, Assoc, AstKind, BinOp, Block, Case, Condition,
    ConditionalBlock, Constant, Enum, EnumVariant, Expr, Field, Function, FunctionBody, Import,
    Item, ItemDecl, ItemQualifiers, LetCondition, Module, Param, ParamQualifiers, Path, Pattern,
    QueryResult, Stmt, StrPart, Type, TypeParam, UnOp, Variance, Visibility, WithSpan, Wrapper,
};
pub use files::{File, SourceLoc, SourceMap};
pub use span::{FileId, Span};
pub use visitor::{AstNode, AstVisitor, NodeId};

pub type Spanned<A> = (A, Span);

pub type SourceAggregate<'src> = Aggregate<'src, WithSpan>;
pub type SourceAnnotation<'src> = Annotation<'src, WithSpan>;
pub type SourceBlock<'src> = Block<'src, WithSpan>;
pub type SourceConditionalBlock<'src> = ConditionalBlock<'src, WithSpan>;
pub type SourceCase<'src> = Case<'src, WithSpan>;
pub type SourcePattern<'src> = Pattern<'src, WithSpan>;
pub type SourceEnum<'src> = Enum<'src, WithSpan>;
pub type SourceExpr<'src> = Expr<'src, WithSpan>;
pub type SourceField<'src> = Field<'src, WithSpan>;
pub type SourceFunction<'src> = Function<'src, WithSpan>;
pub type SourceItem<'src> = Item<'src, WithSpan>;
pub type SourceItemDecl<'src> = ItemDecl<'src, WithSpan>;
pub type SourceModule<'src> = Module<'src, WithSpan>;
pub type SourceParam<'src> = Param<'src, WithSpan>;
pub type SourceStmt<'src> = Stmt<'src, WithSpan>;
pub type SourceType<'src> = Type<'src, WithSpan>;
pub type SourceTypeParam<'src> = TypeParam<'src, WithSpan>;
pub type SourceFunctionBody<'src> = FunctionBody<'src, WithSpan>;
pub type SourceAstNode<'a, 'src> = AstNode<'a, 'src, WithSpan>;
pub type SourceStrPart<'src> = StrPart<'src, WithSpan>;
