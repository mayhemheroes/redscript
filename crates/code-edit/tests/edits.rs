use std::{env, fs};

use redscript_code_edit::{CodeEdit, TextEdit};
use redscript_compiler_api::ast::SourceMap;
use redscript_compiler_api::{
    CompileErrorReporter, SourceMapExt, Symbols, TypeInterner, infer_from_sources,
};

#[test]
fn code_edits() {
    insta::glob!("data/*.reds", |path| {
        let current = env::current_dir().unwrap().canonicalize().unwrap();
        let relative = path
            .strip_prefix(&current)
            .unwrap()
            .to_string_lossy()
            .replace("\\", "/");
        let sources = SourceMap::from_files(&[relative]).unwrap();
        sources.populate_boot_lib();

        let interner = TypeInterner::default();
        let symbols = Symbols::with_default_types();
        let mut reporter = CompileErrorReporter::default();

        let (_, symbols) = infer_from_sources(&sources, symbols, &mut reporter, &interner);

        let reported = reporter.into_reported();
        let edits = reported
            .iter()
            .filter_map(CodeEdit::from_diagnostic)
            .map(|e| TextEdit::from_code_edit(&e, &symbols, &sources))
            .collect::<anyhow::Result<Vec<_>>>()
            .unwrap();

        let mut body = fs::read_to_string(path).unwrap();
        TextEdit::apply_many(edits, &mut body);
        insta::assert_snapshot!(body);
    });
}
