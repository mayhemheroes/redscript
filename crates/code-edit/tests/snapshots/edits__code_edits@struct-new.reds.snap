---
source: crates/code-edit/tests/edits.rs
expression: body
input_file: crates/code-edit/tests/data/struct-new.reds
---
struct Dummy {
  let a: Int32;
  let b: String;
}

struct Tuple<A, B> {
  let first: A;
  let second: B;
}

func Test() {
  let a = Dummy(1, "a");
  let b = Dummy(2, "b");

  let c = Tuple<Int32, String>(3, "c");
  let d = Tuple(4, "d");
}
